src/map.rs

```diff
@@ -945,6 +945,26 @@ where
     /// The key may be any borrowed form of the map's key type, but `Hash` and `Eq` on the borrowed
     /// form must match those for the key type.
     pub fn remove<'g, Q>(&'g self, key: &Q, guard: &'g Guard) -> Option<&'g V>
+        where
+            K: Borrow<Q>,
+            Q: ?Sized + Hash + Eq,
+    {
+        self.replace_node(key, None, None, guard)
+    }
+
+    /// Replaces node value with v, conditional upon match of cv.
+    /// If resulting value does not exists it removes the key (and its corresponding value) from this map.
+    /// This method does nothing if the key is not in the map.
+    /// Returns the previous value associated with the given key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but `Hash` and `Eq` on the borrowed
+    /// form must match those for the key type.
+    pub fn replace_node<'g, Q>(
+        &'g self,
+        key: &Q,
```

Comment:
IMHO, the signature of this function is getting complicated with `new_value` and `old_value`(trying to remember which of the *old* and the *new* comes first), especially considering this function is part of the public API. Should we embed that information in a struct, to make it clearer? Say
```rust
struct NewOldValue {
  new: Option<V>,
  old: Option<Shared<'g, V>>,
}

pub fn replace_node<'g, Q>(&'g self, key: &Q, values: NewOldValue, guard: &'g Guard) -> ...
```

Reply:
`replace_node` should not be public actually and will be change during the PR. Sorry for that.
This said, I'm down for that signature also.

Reply:
I would like it to match the Java version pretty closely so that future changes are easy to adopt to. I think we may want to call it `observed_value` or something like that to communicate intent over just "old".